viper/src/cache.rs

```diff
@@ -0,0 +1,126 @@
+// © 2021, ETH Zurich
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+use serde::{Deserialize, Serialize};
+use std::{
+    collections::HashMap,
+    fs, io,
+    ops::DerefMut,
+    path::{Path, PathBuf},
+    sync::{Arc, Mutex},
+};
+use verification_result::VerificationResult;
+
+pub trait Cache {
+    fn get(&self, request: u64) -> Option<VerificationResult>;
+    fn insert(self, request: u64, result: VerificationResult) -> Option<VerificationResult>;
+}
+// We can change the representation here (e.g. adding fields):
+#[derive(Debug, Clone)]
+pub struct PersistentCache {
+    updated: bool,
+    load_loc: PathBuf,
+    data: HashMap<u64, VerificationResult>,
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+enum ResultCache {
+    V1(HashMap<u64, VerificationResult>),
+    // To save/load different data (e.g. updated PersistentCache) use:
+    // V2(???),
+}
+
+impl From<(PathBuf, ResultCache)> for PersistentCache {
+    /// Used when loading cache from disk
+    fn from((load_loc, rc): (PathBuf, ResultCache)) -> Self {
+        let data = match rc {
+            ResultCache::V1(data) => data,
+            // ResultCache::V2(...) => ...
+        };
+        PersistentCache {
+            updated: false,
+            load_loc,
+            data,
+        }
+    }
+}
+impl From<&PersistentCache> for ResultCache {
+    /// Used when saving cache to disk
+    fn from(cache: &PersistentCache) -> Self {
+        // ResultCache::V2(...)
+        ResultCache::V1(cache.data.clone())
+    }
+}
+
+impl PersistentCache {
+    pub fn load_cache(cache_loc: PathBuf) -> Self {
+        let mut data_res: Option<ResultCache> = None;
+        if !cache_loc.as_os_str().is_empty() {
+            if let Ok(f) = fs::File::open(&cache_loc) {
+                match bincode::deserialize_from(&mut io::BufReader::new(f)) {
+                    Ok(data) => {
+                        info!("Loaded cache from \"{}\"", cache_loc.display());
+                        data_res = Some(data);
+                    }
+                    Err(e) =>
+                        error!("Failed to read cache from \"{}\": {e}", cache_loc.display()),
```

Comment:
Small nit: if the file doesn't exist we can omit this error.

Reply:
Oh, but then `open` fails. Ignore my comment.

Reply:
Yeah, I had a more detailed structure where the first two `if`s would produce `info!` output in the `else` cases but then removed that in order to simplify things and instead just use the generic `info!("Cache file doesn't exist or is invalid. Using fresh cache.")` thing at the end.